dpctl/tensor/libtensor/include/kernels/elementwise_functions/common.hpp

```diff
@@ -0,0 +1,550 @@
+// limitations under the License.
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file defines kernels for elementwise operations over tensor .
+//===----------------------------------------------------------------------===//
+
+#pragma once
+#include <CL/sycl.hpp>
+#include <cstddef>
+#include <cstdint>
+#include <pybind11/pybind11.h>
+
+namespace dpctl
+{
+namespace tensor
+{
+namespace kernels
+{
+namespace elementwise_common
+{
+
+/*! @brief Functor for unary function evaluation on contiguous array */
+template <typename argT,
+          typename resT,
+          typename UnaryOperatorT,
+          unsigned int vec_sz = 4,
+          unsigned int n_vecs = 2>
+struct UnaryContigFunctor
+{
+private:
+    const argT *in = nullptr;
+    resT *out = nullptr;
+    const size_t nelems_;
+
+public:
+    UnaryContigFunctor(const argT *inp, resT *res, const size_t n_elems)
+        : in(inp), out(res), nelems_(n_elems)
+    {
+    }
+
+    void operator()(sycl::nd_item<1> ndit) const
+    {
+        UnaryOperatorT op{};
+        /* Each work-item processes vec_sz elements, contiguous in memory */
+        /* NOTE: vec_sz must divide sg.max_local_range()[0] */
+
+        if constexpr (UnaryOperatorT::is_constant::value) {
+            // value of operator is known to be a known constant
+            constexpr resT const_val = UnaryOperatorT::constant_value;
+            using out_ptrT =
+                sycl::multi_ptr<resT,
+                                sycl::access::address_space::global_space>;
+
+            auto sg = ndit.get_sub_group();
+            std::uint8_t sgSize = sg.get_local_range()[0];
+            std::uint8_t max_sgSize = sg.get_max_local_range()[0];
+            size_t base = n_vecs * vec_sz *
+                          (ndit.get_group(0) * ndit.get_local_range(0) +
+                           sg.get_group_id()[0] * sgSize);
+            if (base + n_vecs * vec_sz * sgSize < nelems_ &&
+                max_sgSize == sgSize) {
+                sycl::vec<resT, vec_sz> res_vec(const_val);
+#pragma unroll
+                for (std::uint8_t it = 0; it < n_vecs * vec_sz; it += vec_sz) {
+                    sg.store<vec_sz>(out_ptrT(&out[base + it * sgSize]),
+                                     res_vec);
+                }
+            }
+            else {
+                for (size_t k = base + sg.get_local_id()[0]; k < nelems_;
+                     k += sgSize) {
+                    out[k] = const_val;
+                }
+            }
+        }
+        else if constexpr (UnaryOperatorT::supports_sg_loadstore::value &&
+                           UnaryOperatorT::supports_vec::value)
+        {
+            using in_ptrT =
+                sycl::multi_ptr<const argT,
+                                sycl::access::address_space::global_space>;
+            using out_ptrT =
+                sycl::multi_ptr<bool,
```

Comment:
Sycl spec only supports relational functions on sycl::vec, so this works fine, but it might be worth adding a comment here that this is why bool is appropriate, or changing it to resT to be consistent.

Reply:
Great catch! I am changing it to `resT`.